pallene/reaching_definitions.lua

```diff
@@ -0,0 +1,497 @@
+
+local ir = require "pallene.ir"
+local location = require "pallene.location"
+
+--local inspect = require "inspect"
+--local ppi = function(t,f) return inspect(t,{newline='',process=f}) end
+--local pp = function(t,f) return inspect(t,{process=f}) end
+
+local RD = {}
+-- start set operations -----------
+local Set = {}
+function Set.copy(t1,t2)
+    for k,v in ipairs(t2) do t1[k] = v end
+end
+function Set.deduplicate(t)
+    local hash = {}
+    local res = {}
+    for _,v in ipairs(t) do
+       if not hash[v] then
+           res[#res+1] = v
+           hash[v] = true
+       end
+    end
+    return res
+end
+function Set.union(t1, t2)
+    local ret = {}
+    for _,v in ipairs(t1) do
+        table.insert(ret,v)
+    end
+    for _,v in ipairs(t2) do
+        table.insert(ret,v)
+    end
+    return Set.deduplicate(ret)
+end
+function Set.difference(t1,t2)
+    local ret = {}
+    local hash = {}
+    for _,v in ipairs(t2) do
+       if not hash[v] then
+           hash[v] = true
+       end
+    end
+    for _,v in ipairs(t1) do
+        if not hash[v] then
+            table.insert(ret,v)
+        end
+    end
+    return ret
+end
+function Set.equal(t1,t2)
+    if #t1 ~= #t2 then return false end
+    local h1,h2 = {},{}
+    for k,v in ipairs(t1) do
+       if not h1[v] then
+           h1[v] = true
+       end
+       if not h2[t2[k]] then
+           h2[t2[k]] = true
+       end
+    end
+    for k,_ in pairs(h1) do
+        if not h2[k] then return false end
+    end
+    return true
+end
+
+
+-- end   set operations -----------
+-- start stack operations ---------
+local Stack = {}
+function Stack.push(t,elem)
+    t[#t+1] = elem
+end
+function Stack.top(t)
+    return t[#t]
+end
+function Stack.pop(t)
+    local ret = t[#t]
+    t[#t] = nil
+    return ret
+end
+function Stack.copy(t1,t2)
+    for k,v in ipairs(t2) do t1[k] = v end
+end
+-- end   stack operations ---------
+-- start control flow -------------
+local CFG = {}
+--[[
+    A control flow graph is a representation of all paths 
+    that might be traversed through a program execution. 
+    
+    The nodes represent commands that use or alter data 
+    and the edges represent control instructions.
+
+    In this module, each node is defined by two sets of node indices (from and to)
+    and a list of commands. These commands are pointers to nodes in the ir module.
+
+    All the commands of the list are one of the listed in the ir module
+    except for the control flow ones (Nop, Seq, Return, Break, Loop, If and For).
+
+    Values aren't nodes.
+
+--]]
+function CFG:new(code)
+    local new = {}
+    setmetatable(new, self)
+    self.__index = self
+    -- exit node
+    new[-1] = {cmd = false, to={},from={}}
+    -- entry node
+    new[0] = {cmd = false, to={},from={}}
```

Comment:
I've thought of that, but since I was using numerical indexes, the introduction of string indexes could introduce bugs. But I will see if there is any problem with that and change it accordingly.